qiskit/transpiler/passes/optimization/optimize_annotated.py

```diff
@@ -0,0 +1,188 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Optimize annotated operations on a circuit."""
+
+from typing import Optional, List, Tuple
+
+from qiskit.converters import circuit_to_dag, dag_to_circuit
+from qiskit.circuit.annotated_operation import AnnotatedOperation, _canonicalize_modifiers
+from qiskit.circuit import EquivalenceLibrary, ControlledGate, Operation, ControlFlowOp
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.passes.utils import control_flow
+from qiskit.transpiler.target import Target
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.transpiler.exceptions import TranspilerError
+
+
+class OptimizeAnnotated(TransformationPass):
+    """Optimization pass on circuits with annotated operations."""
+
+    def __init__(
+        self,
+        target: Optional[Target] = None,
+        equivalence_library: Optional[EquivalenceLibrary] = None,
+        basis_gates: Optional[List[str]] = None,
+    ):
+        """
+        OptimizeAnnotated initializer.
+
+        Args:
+            target: Optional, the backend target to use for this pass.
+            equivalence_library: The equivalence library used
+                (instructions in this library will not be optimized by this pass).
+            basis_gates: Optional, target basis names to unroll to, e.g. `['u3', 'cx']`
+                (instructions in this list will not be optimized by this pass).
+                Ignored if ``target`` is also specified.
+        """
+        super().__init__()
+
+        self._target = target
+        self._equiv_lib = equivalence_library
+        self._basis_gates = basis_gates
+
+        self._top_level_only = self._basis_gates is None and self._target is None
+
+        if not self._top_level_only and self._target is None:
+            basic_insts = {"measure", "reset", "barrier", "snapshot", "delay"}
+            self._device_insts = basic_insts | set(self._basis_gates)
+
+    def run(self, dag: DAGCircuit):
+        """Run the OptimizeAnnotated pass on `dag`.
+
+        Args:
+            dag: input dag.
+
+        Returns:
+            Output dag with higher-level operations optimized.
+
+        Raises:
+            TranspilerError: when something goes wrong.
+
+        """
+        dag, _ = self._run_inner(dag)
+        return dag
+
+    def _run_inner(self, dag) -> Tuple[DAGCircuit, bool]:
+        """
+        Optimizes annotated operations.
+        Returns True if did something.
+        """
+        # Handle control-flow
```

Comment:
Do you think we should have a fast return here if there are no annotated ops or control flow ops in the circuit? Something like:

```suggestion
        op_names = dag.count_ops(recurse=False)
        if "annotated" not in op_names and not CONTROL_FLOW_OP_NAMES.intersection(op_names):
            return dag, False
        # Handle control-flow
```
if

Reply:
Good suggestion, added in [1b390e5](https://github.com/Qiskit/qiskit/pull/11476/commits/1b390e5bc6813967bd7f8054cccfb054f9c7fc25).